app/src/pages/playground/PlaygroundOutput.tsx

```diff
@@ -67,80 +104,367 @@ function PlaygroundOutputMessage({ message }: { message: ChatMessage }) {
+        });
+        return;
+      }
+      if (chatCompletion.__typename === "FinishedChatCompletion") {
+        updateInstance({
+          instanceId,
+          patch: {
+            spanId: chatCompletion.span.id,
+          },
+        });
+        if (chatCompletion.errorMessage != null) {
+          notifyError({
+            title: "Chat completion failed",
+            message: chatCompletion.errorMessage,
+            expireMs: 10000,
+          });
+        }
+      }
+    },
+    [
+      instanceId,
+      markPlaygroundInstanceComplete,
+      notifyError,
+      props.playgroundInstanceId,
+      updateInstance,
+    ]
+  );
+
+  const startStreaming = useChatCompletionSubscription({
+    params: chatCompletionParams,
+    onNext,
+    onCompleted: () => {
+      markPlaygroundInstanceComplete(props.playgroundInstanceId);
+    },
+    onFailed: (error) => {
+      // TODO(apowell): We should display this error to the user after formatting it nicely.
+      // eslint-disable-next-line no-console
+      console.error(error);
+      markPlaygroundInstanceComplete(props.playgroundInstanceId);
+      updateInstance({
+        instanceId: props.playgroundInstanceId,
+        patch: {
+          activeRunId: null,
+        },
+      });
+      notifyError({
+        title: "Failed to get output",
+        message: "Please try again.",
+      });
+    },
+  });
 
-      return messages.map((message, index) => {
-        return <PlaygroundOutputMessage key={index} message={message} />;
+  const onCompleted = useCallback(
+    (
+      response: PlaygroundOutputMutation$data,
+      errors: PayloadError[] | null
+    ) => {
+      markPlaygroundInstanceComplete(props.playgroundInstanceId);
+      updateInstance({
+        instanceId,
+        patch: {
+          spanId: response.chatCompletion.span.id,
+        },
       });
+      if (errors) {
+        notifyError({
+          title: "Chat completion failed",
+          message: errors[0].message,
+        });
+        return;
+      }
+      if (response.chatCompletion.errorMessage != null) {
+        notifyError({
+          title: "Chat completion failed",
+          message: response.chatCompletion.errorMessage,
+        });
+        return;
+      }
+      setOutputContent(response.chatCompletion.content ?? undefined);
+      setToolCalls(response.chatCompletion.toolCalls);
+    },
+    [
+      instanceId,
+      markPlaygroundInstanceComplete,
+      notifyError,
+      props.playgroundInstanceId,
+      updateInstance,
+    ]
+  );
+
+  useEffect(() => {
+    if (!hasRunId) {
+      return;
     }
-    if (typeof instance.output === "string") {
-      return (
-        <PlaygroundOutputMessage
-          message={{
-            id: generateMessageId(),
-            content: instance.output,
-            role: "ai",
-          }}
-        />
-      );
+    setOutputContent(undefined);
+    setToolCalls([]);
+
+    if (streaming) {
+      startStreaming();
+      return;
```

Comment:
Should this return the disposal function from start streaming? That way we clean up any lingering streams when we start a new one

Reply:
that's returned by the useSubscription hook, not sureif it should be done here or there, let me check the docs, any intuition on that?